### PR TITLE
[BEAM-1050] Fixing pipeline's result state in streaming mode

### PR DESCRIPTION
**Pull request by staslev**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

R: @amitsela

### CODE REVIEW

**Comment by staslev** [comment]
Been having troubles to get the build to pass with tests, but the failures seem to be unrelated to the Spark runner module as they originate from the Apex runner tests:

```
testContainsInAnyOrder(org.apache.beam.runners.apex.translation.ParDoBoundTranslatorTest)  Time elapsed: 7.175 sec  <<< ERROR!
java.lang.RuntimeException: Error creating local cluster
```

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java [hunk 1]
```diff
@@ -262,34 +263,52 @@ private State cancel(boolean gracefully) throws IOException {
...
-  @Override
-  public State waitUntilFinish() {
-    return waitUntilFinish(Duration.ZERO);
-  }
-
-  @Override
-  public State waitUntilFinish(Duration duration) {
-    if (isStreamingPipeline()) {
+  private State waitOnStreaming(Duration duration) {
+    final State newState;
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java]: maybe `finalState`  or `finishState` ?

File: runners/spark/src/test/java/org/apache/beam/runners/spark/FailFastOnMicrobatchFailureTest.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import avro.shaded.com.google.common.collect.Lists;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * This suite tests that verifies a Beam over spark pipeline fails fast upon a failed batch and
+ * does not keep on running upon encountering a batch failure.
+ */
+public class FailFastOnMicrobatchFailureTest implements Serializable {
```
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/FailFastOnMicrobatchFailureTest.java]: Let's make this suite about result state in general, having the streaming fail test for now (pending work on batch).
The test itself can be simpler too.

Comment (staslev) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/FailFastOnMicrobatchFailureTest.java]: Sure, I'll make it a bit more straight forward.

Would you like to open a separate issue regarding the consolidation of `PipelineResult.State` tests into this suite so it can be managed independently of this PR?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java [hunk 3]
```diff
@@ -226,7 +226,8 @@ public MetricResults metrics() {
   @Override
   public void close(boolean gracefully) {
     // Stopping streaming job if running
-    if (isStreamingPipeline() && !state.isTerminal()) {
+    // state can be null in case the pipeline has timed out
+    if (isStreamingPipeline() && state != null && !state.isTerminal()) {
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java]: 👍

File: runners/spark/src/test/java/org/apache/beam/runners/spark/FailFastOnMicrobatchFailureTest.java (hunk 2 above)
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/FailFastOnMicrobatchFailureTest.java]: I don't think so.
`Done` is mostly straight forward, and can be asserted on any existing batch/streaming test, fail states are (evidently) less straight forward, and since we made modifications to properly handle them, I think they deserve their own tests (again, not that I'm suggesting not to test "positive-path" but we can piggyback on top of existing tests).

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SideEffectsTest.java [hunk 4]
```diff
@@ -44,7 +45,8 @@
 
   @Test
   public void test() throws Exception {
-    Pipeline p = Pipeline.create(pipelineOptions.getOptions());
+    SparkPipelineOptions options = pipelineOptions.getOptions();
```
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SideEffectsTest.java]: Is this really necessary ?

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/EmptyStreamAssertionTest.java [hunk 5]
```diff
@@ -54,6 +55,9 @@
   public SparkTestPipelineOptionsForStreaming commonOptions =
       new SparkTestPipelineOptionsForStreaming();
 
+  @Rule
+  public ClearAggregatorsRule clearAggregatorsRule = new ClearAggregatorsRule();
```
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/EmptyStreamAssertionTest.java]: This seems unused, right ?

File: runners/spark/src/test/java/org/apache/beam/runners/spark/aggregators/metrics/sink/ClearAggregatorsRule.java [hunk 6]
```diff
@@ -25,7 +25,7 @@
  * A rule that clears the {@link org.apache.beam.runners.spark.aggregators.AccumulatorSingleton}
  * which represents the Beam {@link org.apache.beam.sdk.transforms.Aggregator}s.
  */
-class ClearAggregatorsRule extends ExternalResource {
+public class ClearAggregatorsRule extends ExternalResource {
```
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/aggregators/metrics/sink/ClearAggregatorsRule.java]: Why is this necessary ?

**Comment by amitsela** [comment]
@staslev see comments, thanks!
Jenkins complains on import order issues in `FailFastOnMicrobatchFailureTest`, check it out.

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SideEffectsTest.java (hunk 4 above)
Comment (staslev) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SideEffectsTest.java]: Nope, must be something I have forgot to change back after experimenting.

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/EmptyStreamAssertionTest.java (hunk 5 above)
Comment (staslev) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/EmptyStreamAssertionTest.java]: `PAssertStreaming#runAndAssertContents` asserts aggregators, so this rule is here to make sure they are cleared before running the tests.

I actually added this rule here after seeing some failures stemming from stale aggregator values.

File: runners/spark/src/test/java/org/apache/beam/runners/spark/FailFastOnMicrobatchFailureTest.java (hunk 2 above)
Comment (staslev) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/FailFastOnMicrobatchFailureTest.java]: My suggestion to make it a separate PR was not implying these scenarios did not deserve to be tested. It was aimed at task & time management. The work done in this PR could be merged now (after we're done reviewing), while the additional work you describe can be carried out at a later time. Perhaps
when we make the changes that align the batch and streaming modes in `SparkRunner#run`, and make `PipelineResult#waitUntilFinish` behave in a more natural way in `SparkRunner`.

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/EmptyStreamAssertionTest.java (hunk 5 above)
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/EmptyStreamAssertionTest.java]: 👍

File: runners/spark/src/test/java/org/apache/beam/runners/spark/aggregators/metrics/sink/ClearAggregatorsRule.java (hunk 6 above)
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/aggregators/metrics/sink/ClearAggregatorsRule.java]: 👍

**Comment by staslev** [comment]
While I initially thought I wouldn't be able get around to addressing the underlying issues in `SparkRunner`, `EvaluationContext` and the likes, I ended up giving it a try anyway.

I have another branch based on this PR and all the changes from this review, so let's close this one and I'll make a new PR based on the new branch. Sounds good?

**Comment by amitsela** [comment]
Sure, only you can close this PR though 😉

**Comment by staslev** [comment]
Closing, moved to #1466.
